src/main/java/uk/nhs/tis/trainee/notifications/migration/InsertScheduledEmailHistory.java

```diff
@@ -61,14 +76,34 @@ public void migrate() throws SchedulerException {
     for (String groupName : scheduler.getJobGroupNames()) {
       for (JobKey jobKey : scheduler.getJobKeys(GroupMatcher.jobGroupEquals(groupName))) {
 
-        // Get scheduled job
+        // Get scheduled job from Quartz
         List<Trigger> triggers = (List<Trigger>) scheduler.getTriggersOfJob(jobKey);
         String jobName = jobKey.getName();
         Date when = triggers.get(0).getNextFireTime();
         JobDataMap jobDetails = scheduler.getJobDetail(jobKey).getJobDataMap();
-        log.info("Processing scheduled email from Quartz: [jobName] : " + jobName + " - " + when);
 
-        notificationService.saveScheduleHistory(jobDetails, when);
+        // Check if it's already migrated to DB
+        History scheduledHistory = null;
+        NotificationType notificationType =
+            NotificationType.valueOf(jobDetails.get(TEMPLATE_NOTIFICATION_TYPE_FIELD).toString());
+        History.TisReferenceInfo tisReferenceInfo =
+            getTisReferenceInfo(jobDetails, notificationType);
+        String personId = jobDetails.getString(PERSON_ID_FIELD);
+        if (tisReferenceInfo != null) {
```

Comment:
This seems ok, but I wonder if iterating through all the (~85k) jobs in quartz and checking if each is in mongo is going to happen in a reasonable amount of time? On preprod there are ~29k, so maybe that is enough to get a sense if it can complete before it gets timed-out.
The alternative might be to delete the quartz triggers as they are added to the DB so they won't be reprocessed, but I'm happy to try your way first 😁

Reply:
(Sorry, this comment should have been under line 96)

Reply:
We cannot delete the quartz trigger at this point I am afraid, as I haven't work that far to refractor to trigger send email depends on the DB records, so the quartz is still need for trigger.
I am going to add a log after al the for loops for notify the end of migration. Is there any other suggestion I can add to improve it?

Reply:
Ah ok. Seems good to me - let's give it a try!